### DISCUSSION TITLE
@types/[node] - 'ReadableByteStreamController' is referenced directly or indirectly in its own type annotation.

### DISCUSSION THREAD

**Discussion by navneerajmishra** [DefinitelyTyped/DefinitelyTyped#70432] [Issues with a @types package]
The build gets failed since yesterday with following error
Error: node_modules/@types/node/stream/web.d.ts:484:13 - error TS2502: 'ReadableByteStreamController' is referenced directly or indirectly in its own type annotation.

![image](https://github.com/user-attachments/assets/e0c5527a-6e4d-4c8f-9562-74e1b7e72804)

Looks like it happened after merging this PR - https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70269/files
Can someone please help with this ASAP.

[a comment by an automated account is not shown]

**Comment by ThomasNB** [comment]
I have the same problem with an additional ERROR:

```
14:08:41   ERROR  Failed to compile with 4 errors2:08:32 PM
14:08:41  
14:08:41   error  in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts
14:08:41  
14:08:41  ERROR in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts(469,56):
14:08:41  469:56 '?' expected.
14:08:41      467 |             // If ReportingObserver check is removed, the type here will form a circular reference in TS5.0+lib.dom.d.ts
14:08:41      468 |             ReportingObserver: any;
14:08:41    > 469 |             DecompressionStream: infer T extends object;
14:08:41          |                                                        ^
14:08:41      470 |         } ? T
14:08:41      471 |             // TS 4.8, 4.9, 5.0
14:08:41      472 |             : typeof globalThis extends { onmessage: any; TransformStream: { prototype: infer T } } ? {
14:08:41  
14:08:41   error  in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts
14:08:41  
14:08:41  ERROR in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts(484,13):
14:08:41  484:13 'ReadableByteStreamController' is referenced directly or indirectly in its own type annotation.
14:08:41      482 |          * @since v18.0.0
14:08:41      483 |          */
14:08:41    > 484 |         var ReadableByteStreamController: typeof globalThis extends
14:08:41          |             ^
14:08:41      485 |             { onmessage: any; ReadableByteStreamController: infer T } ? T
14:08:41      486 |             : typeof import("stream/web").ReadableByteStreamController;
14:08:41      487 | 
14:08:41  
14:08:41   error  in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts
14:08:41  
14:08:41  ERROR in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts(503,13):
14:08:41  503:13 'ReadableStreamBYOBReader' is referenced directly or indirectly in its own type annotation.
14:08:41      501 |          * @since v18.0.0
14:08:41      502 |          */
14:08:41    > 503 |         var ReadableStreamBYOBReader: typeof globalThis extends { onmessage: any; ReadableStreamBYOBReader: infer T }
14:08:41          |             ^
14:08:41      504 |             ? T
14:08:41      505 |             : typeof import("stream/web").ReadableStreamBYOBReader;
14:08:41      506 | 
14:08:41  
14:08:41   error  in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts
14:08:41  
14:08:41  ERROR in /home/jenkins/workspace/workspace/tionsmodul_slutafregning_MR-1499/client/node_modules/@types/node/stream/web.d.ts(513,13):
14:08:41  513:13 'ReadableStreamBYOBRequest' is referenced directly or indirectly in its own type annotation.
14:08:41      511 |          * @since v18.0.0
14:08:41      512 |          */
14:08:41    > 513 |         var ReadableStreamBYOBRequest: typeof globalThis extends { onmessage: any; ReadableStreamBYOBRequest: infer T }
14:08:41          |             ^
14:08:41      514 |             ? T
14:08:41      515 |             : typeof import("stream/web").ReadableStreamBYOBRequest;
14:08:41      516 | 
```

**Reply by navneerajmishra** [reply]
@ikeyan - Could you please help us out with this?

**Reply by Renegade334** [reply]
@navneerajmishra et al. can you please confirm what version of TS you're building against, and share your tsconfig?

**Reply by navneerajmishra** [reply]
@Renegade334 "typescript": "~4.7.2"
[tsconfig.json](https://github.com/user-attachments/files/16884942/tsconfig.json)

**Reply by LasseRB** [reply]
@Renegade334 
I'm from the same team as @ThomasNB.
We use typescript version `4.6.4`. 
tsconfig:
```
{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "strict": true,
    "jsx": "preserve",
    "importHelpers": true,
    "moduleResolution": "node",
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "baseUrl": ".",
    "types": ["webpack-env", "jest"],
    "paths": {
      "@/*": ["src/*"],
      "~/*": ["tests/*"]
    },
    "lib": ["esnext", "dom", "dom.iterable", "scripthost"]
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.tsx",
    "src/**/*.vue",
    "tests/**/*.ts",
    "tests/**/*.tsx",
    "./babel.config.js",
    "./jest.config.js",
    "./vue.config.js"
  ],
  "exclude": ["node_modules", "src/tests/**"]
}

```

**Reply by ikeyan** [reply]
@navneerajmishra @ThomasNB @LasseRB 
from [README](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#support-window)

> Definitely Typed only tests packages on versions of TypeScript that are less than 2 years old.

So TS4.7 or older are not supported now.
If you use older TS, you should depend on specific tags like `"@types/node": "ts4.7"`.

https://www.npmjs.com/package/@types/node?activeTab=versions

**Comment by RobertCraigie** [comment]
We're also seeing this error

**Reply by RobertCraigie** [reply]
You can reproduce this with:
```bash
git clone --depth 1 https://github.com/openai/openai-node --branch before-node-types-fix
cd openai-node
yarn
yarn tsn ecosystem-tests/cli.ts node-ts-cjs
```
I've [worked around](https://github.com/openai/openai-node/commit/3c7bdfdb373bff77db0e3fecd5d49ddfa4284cd9) this by pinning `@types/node` as a transitive dep.

**Reply by ikeyan** [reply]
@RobertCraigie I was asked for OPENAI_API_KEY, so I modified the 4th line to 
```
OPENAI_API_KEY=A yarn tsn ecosystem-tests/cli.ts --verbose node-ts-cjs
```
Looks like typescript 4.7.4 is used, which is out of support.

After I replaced `tsc` in `ecosystem-tests/node-ts-cjs/package.json` to `../../node_modules/.bin/tsc` (4.9.5), node-ts-cjs compiled successfully.

**Reply by RobertCraigie** [reply]
Ah I didn't realise 4.7.4 wasn't supported, when was support dropped?

**Reply by ikeyan** [reply]
> Definitely Typed only tests packages on versions of TypeScript that are less than 2 years old.

from [README](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#support-window)

If you use older TS, you should depend on specific tags like `"@types/node": "ts4.7"`.

https://www.npmjs.com/package/@types/node?activeTab=versions

**Comment by ThomasNB** [comment]
A temporary work around can be adding this to `package.json`:

```
"overrides": {
  "@types/node": "22.5.1"
}
```

I have seen this work for `npm` - for `yarn` you should be able to use `resolutions` instead.

**Reply by navneerajmishra** [reply]
Yes, that's the workaround we've opted for as well.

**Comment by meszarosgabor12** [comment]
Same problem here :/

```
node_modules/@types/node/stream/web.d.ts:484:13 - error TS2502: 'ReadableByteStreamController' is referenced directly or indirectly in its own type annotation.

484         var ReadableByteStreamController: typeof globalThis extends
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
node_modules/@types/node/stream/web.d.ts:503:13 - error TS2502: 'ReadableStreamBYOBReader' is referenced directly or indirectly in its own type annotation.

503         var ReadableStreamBYOBReader: typeof globalThis extends { onmessage: any; ReadableStreamBYOBReader: infer T }
                ~~~~~~~~~~~~~~~~~~~~~~~~
node_modules/@types/node/stream/web.d.ts:513:13 - error TS2502: 'ReadableStreamBYOBRequest' is referenced directly or indirectly in its own type annotation.

513         var ReadableStreamBYOBRequest: typeof globalThis extends { onmessage: any; ReadableStreamBYOBRequest: infer T }
                ~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Reply by ikeyan** [reply]
@meszarosgabor12 In case TS4.7 is used:
from [README](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#support-window)

> Definitely Typed only tests packages on versions of TypeScript that are less than 2 years old.

So TS4.7 or older are not supported now.
If you use older TS, you should depend on specific tags like `"@types/node": "ts4.7"`.

https://www.npmjs.com/package/@types/node?activeTab=versions

**Comment by Renegade334** [comment]
Just putting a top-level answer in here for visibility, although it's essentially the same as @ikeyan's comments.

These library compilation errors will only appear on TS 4.7.4 or earlier, as the DOM interfaces in the inference checks were added to the TS libraries in 4.8.

DefinitelyTyped packages only support TypeScript versions released within the last two years ([support matrix](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#support-window)), which makes TS 4.9 the earliest supported version at present.

If you're running into this error, you have a couple of options:

#### 1) Constrain your @types/node dependency to a supported version
The latest formally-supported @types/node versions are 20.11.25 for TS 4.6 and 20.14.8 for TS 4.7. Constraining your @types/node dependency to a supported version will fix the lib check errors during compilation and should avoid breakage going forward.

DefinitelyTyped also exports the ["tsX.X" package tags](https://www.npmjs.com/package/@types/node?activeTab=versions#tabpanel-versions) to facilitate this, so this can also be used to constrain your dependencies:
```json
"devDependencies": {
  "typescript": "~4.7.0",
  "@types/node": "ts4.7"
}
```

#### 2) Update your TypeScript dependency
Updating to a TypeScript version within DT support will also resolve the compilation errors. Note, however, that if you're constraining your TypeScript dependency, it's also good practice to constrain @types/node to avoid this eventuality in the future.
